### ISSUE TITLE
Nullable state tracking for the ternary operator

### ISSUE THREAD

**Issue by svick** [dotnet/roslyn#44498]
Consider the following code:

```c#
if (response != null ? response.IsSuccessStatusCode : false)
{
    _ = response.Content;
}
```

[This currently produces "warning CS8602: Dereference of a possibly null reference"](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgAwAEhAjAHQBycMVAEjDAA4DcAsAFA8DEAOwCuAGxERgIuKTgCJUnj0IAmUgGEeAbx6ld5ACykAsgApmbAEpworAPYCocI9agQA5tITW7DuAEodPW1uPVDSHFITLxt7R1IAQgBeUmExUgB+UmifRyoASSgAZSEAYxKXQpgIGCEoNVsAE2kQUmwIEUcAkLDdYJ6egH1SZOzYuCp6gXgprm6egF9A3UXueaA=) in the body of the `if`, even though `response` can't possibly be `null` there.

This code can be simplified to the following:

```c#
if (response != null && response.IsSuccessStatusCode)
{
    _ = response.Content;
}
```

[This version does not produce any warnings.](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgAwAEhAjAHQBycMVAEjDAA4DcAsAFA8DEAOwCuAGxERgIuKTgCJUnj0IAmUgGEeAbx6ld5ACykAsgApmbAEpworAPYCocI9agQA5tITW7DuAEodPW1uPVDSHFITLxt7R1IAQgBeUmExUgAydNJon0cqAEkoAGUhAGNSlyKYCBghKDVbABN/QLDSYLa2gH1SZJzYuCoGgXgRrhDOgF9W0mnuSaA=)

So, my question is: should nullability tracking be changed so that it also works for the ternary operator in situations like the above (i.e. where one of the possible results is a constant boolean)?

One argument for why that's not necessary is that all such cases can (and should?) be simplified to use `&&` or `||`, which track nullability properly. But personally, I think the language should work well even for code that's still being written, before it's refactored to its final form, so it makes sense to support this.

**Comment by RikkiGibson** [comment]
I see no reason why we shouldn't do this for nullable, but doing it for some other forms of analysis (such as definite assignment of 'out' arguments described in #44471) may require spec changes.

**Comment by RikkiGibson** [comment]
This will be resolved by the "improved definite assignment" feature in C# 10. dotnet/csharplang#4465 #51463
